src/Workspaces/Core/Portable/Workspace/Host/DocumentService/ISpanMapperService.cs

```diff
@@ -0,0 +1,31 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Text;
+
+namespace Microsoft.CodeAnalysis.Host
+{
+    /// <summary>
+    /// Map spans in a document to other spans even in other document
+    ///
+    /// this will be used by various features if provided to convert span in one document to other spans.
+    /// 
+    /// for example, it is used to show spans users expect in a razor file rather than spans in 
+    /// auto generated file that is implementation detail or navigate to the right place rather 
+    /// than the generated file and etc.
+    /// </summary>
+    internal interface ISpanMapperService : IDocumentService
+    {
+        /// <summary>
+        /// Map spans in the document to more appropriate location for this document
+        /// </summary>
+        /// <param name="document">Document given spans belong to</param>
+        /// <param name="spans">Spans in the document</param>
+        /// <param name="cancellationToken">Cancellation token</param>
+        /// <returns>Return mapped span. order of result should be same as the given span</returns>
+        Task<ImmutableArray<DocumentSpan>> MapSpansAsync(Document document, IEnumerable<TextSpan> spans, CancellationToken cancellationToken);
```

Comment:
Commented on the implementation, but it doesn't seem that DocumentSpan is the right currency here. In the case we're trying to address, there's a .cshtml file on disk, and we have a file being added for that source. But we _don't_ actually have a document _for the .cshtml file itself_.

Reply:
the work Ryan is doing is adding Document for all closed files.

Reply:
> But we don't actually have a document for the .cshtml file itself.

Right, I agree with @jasonmalinowski - you guys will have a `Document` that contains the generated code (C#), but not a document that contains the original source code (.cshtml)